### ISSUE TITLE
Failing Health Checks Only When Website is Made Live

### ISSUE THREAD

**Issue by ILLUMINICE** [widdix/aws-cf-templates#110]
I've run into a very strange and difficult to troubleshoot issue tonight that is causing my instances to fail health checks one after the next. The instances don't last two minutes before failing, and then so on in a never-ending loop of failing instances.

This all started earlier when I turned off a password protection plugin that I'm using to keep our site private during development. All the plugin does is create a password page as a substitute for the normal WordPress index page. It's very much like an htaccess password, just using a plugin instead. So I turned this off temporarily while doing some work that necessitated the home page being accessible. That's when the problems began.

For some reason when the website is completely open (without that password plugin), the health checks start to fail. I'm guessing this issue has been there a while, and I simply didn't know until now (because the site was password protected).

So I started digging into the system logs of the failed instances and found this error occurring immediately before they would shut down:

**login: /dev/fd/11: line 1: /sbin/plymouthd: No such file or directory
initctl: Event failed**

After this you can see the services start shutting down on the instance. As far as I can tell there are no other errors occurring, at least in the system log.

I'm also able to see a bunch of failures in this log related to the failing instances: /var/log/cloud-init.log

I'm really at a loss as to what is going on here, and hope someone can help shed light on this.

~ Michael

**Comment by ILLUMINICE** [comment]
I wanted to post an update here that whatever is causing these failures is unrelated to WordPress plugins or the CDN.

I have deactivate all WordPress plugins to rule theme out. I've also tried bypassing the CDN and pointing directly towards the loan balancer, which has had no effect.

I've taken a good a look at my htaccess file to ensure there was nothing in there that could cause this too.

.....

If I make the site public, the instances start to fail immediately. If I make the site private (password protecting it), then the instances stabilize.

I can't imagine what is causing this, and really could use some insight from anyone here...

~ Michael

**Comment by andreaswittig** [comment]
I guess the health check of the load balancer is failing.

1. Make your site public.
1. SSH into one of your EC2 Instances. Run the following comman:

```
curl -v http://localhost:80/
```

What's the returned HTTP status code?

**Comment by ILLUMINICE** [comment]
Hi Andreas,

Your timing is impeccable! I've been trying to get to the bottom of this issue all day and am starting to think it's resource related, because I've been using the default t2.micro instance. I've read many things about these micro instances being unstable and barely able to handle WordPress. Frankly I left it this way originally only because we're in development and I'm the only person using the site.

What I can't really reconcile here is why the server is stable when I have the website password protected, and then it suddenly dies when I make the site public. This makes no sense to me since the website is doing the exact same thing once I am inside it and past the password.

I was about to upgrade both the EC2 and RDS instance types when I saw your message. But I'm going to try your suggestion above first on the existing instance type to see what happens.

Btw, I've also looked at the memory usage, and it's within 100-130 MB of max when nothing at all is occurring on the site. They really mean it when they call it a micro instance!

I'll message back here momentarily after trying your idea...

~ Michael

**Comment by ILLUMINICE** [comment]
Ok, I made the site public, and naturally the instances have started crashing. I ran that command and it's bringing up a lot of information. It looks like an entire web page here. Where exactly am I supposed to see the status code?

**Comment by ILLUMINICE** [comment]
These are the headers:

* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
[ec2-user@ip-10-32-40-150 ~]$ curl -v http://localhost:80/
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.51.0
> Accept: */*
>
< HTTP/1.1 403 Forbidden
< Date: Sat, 26 Aug 2017 09:14:02 GMT
< Server: Apache/2.4.27 (Amazon) PHP/7.0.21
< Last-Modified: Wed, 02 Aug 2017 18:01:11 GMT
< ETag: "131b-555c90c8e67c0"
< Accept-Ranges: bytes
< Content-Length: 4891
< Content-Type: text/html; charset=UTF-8
<
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">

**Comment by ILLUMINICE** [comment]
Ok, I believe the status code is what you see above: 403 Forbidden...

**Comment by ILLUMINICE** [comment]
I'm upgrading my EC2 instance type to m3.medium and the Aurora db to db.t2.medium.

Let's see if this clears things up. If not then we know it's something other than resources in all likelihood.

~ Michael

**Comment by ILLUMINICE** [comment]
The good news is that I was able to successfully upgrade the EC2 and RDS instance types.

The bad news is that this made no difference for the problem. When the site is live the instances start failing health checks and are shut down one after the next. Magically if the site is password protected this doesn't occur, even when browsing the site.

I'm really at a loss here...

You can see what things look like a little in these screenshots. All of these instances have terminated due to health check failures, and it only happens when the website is live. I should mention that the website itself is working perfectly when this occurs. The instances last about three minutes before failing.

The screenshot of the memory usage was taken in the moments before that instance was shut down due to a health check. This definitely is not a memory problem.

![1](https://user-images.githubusercontent.com/11562655/29740616-637d7fa4-8a0f-11e7-8002-99dd540881f3.jpg)
![2](https://user-images.githubusercontent.com/11562655/29740617-637ef2b2-8a0f-11e7-8268-5de18154e794.jpg)
![3](https://user-images.githubusercontent.com/11562655/29740618-637f72b4-8a0f-11e7-886e-1d152d6bb4aa.jpg)

**Comment by andreaswittig** [comment]
> These are the headers:
> 
> Curl_http_done: called premature == 0
> Connection #0 to host localhost left intact
> [ec2-user@ip-10-32-40-150 ~]$ curl -v http://localhost:80/
> Trying 127.0.0.1...
> TCP_NODELAY set
> Connected to localhost (127.0.0.1) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.51.0
> Accept: /
> < HTTP/1.1 403 Forbidden
> < Date: Sat, 26 Aug 2017 09:14:02 GMT
> < Server: Apache/2.4.27 (Amazon) PHP/7.0.21
> < Last-Modified: Wed, 02 Aug 2017 18:01:11 GMT
> < ETag: "131b-555c90c8e67c0"
> < Accept-Ranges: bytes
> < Content-Length: 4891
> < Content-Type: text/html; charset=UTF-8
> <

That's the problem. WordPress returns a 403 error. That's causing the health check of the load balancer to fail.

You could update the health check to accept all status codes from 200-499 to debug the issue.

```
...
LoadBalancerTargetGroup:
    Type: "AWS::ElasticLoadBalancingV2::TargetGroup"
    Properties:
      HealthCheckIntervalSeconds: 5
      HealthCheckPath: '/'
      HealthCheckPort: 80
      HealthCheckProtocol: HTTP
      HealthCheckTimeoutSeconds: 3
      HealthyThresholdCount: 2
      UnhealthyThresholdCount: 2
      Matcher:
        HttpCode: '200-499'
      Port: 80
      Protocol: HTTP
      VpcId:
        'Fn::ImportValue': !Sub '${ParentVPCStack}-VPC'
...
```

**Comment by ILLUMINICE** [comment]
**Ok, I have some big news... Issue #111 and this issue are the same!**

Thank you @gerrymiller for pointing me in the right direction.

For some reason standard WordPress permalink rewrites in htaccess are causing the health checks to fail. I've actually seen failures like this before associated with rewrites in htaccess, though never before from the permalinks.

These are the standard rewrites causing the problem:

# BEGIN WordPress
<IfModule mod_rewrite.c>
RewriteEngine On
RewriteBase /
RewriteRule ^index\.php$ - [L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule . /index.php [L]
</IfModule>
# END WordPress

If you set your permalinks to standard, the above rewrites are removed from htaccess and the health checks pass. If you choose any custom permalinks this code is necessary, and the health checks fail.

The reason my website wasn't having this problem when password protected was because it was preventing the htaccess file from being read. I don't understand why the health checks didn't fail however once I entered my website. None the less, it is the rewrites causing the failures for whatever reason.

There's more to this though...

Upon the recommendation of Andreas I started looking into those HTTP statuses in the headers. I found there is a strange discrepancy between what the server says for the status and what outside status checkers see.

Take a look at my attached screenshot. You can clearly see that when checking the HTTP status at command line it says "403 Forbidden." However when checking from the outside it results in "200 OK." I don't know if this is at all related to the permalinks problem, but it may be, and it is certainly strange. I can also confirm that the HTTP status does not change (and has this discrepancy) whether or not the permalink rewrites are in place.

![headerstatus](https://user-images.githubusercontent.com/11562655/29798612-d9e4fafa-8c12-11e7-9108-a64b5c1f30d1.jpg)

So the question here is why are standard WordPress permalink rewrites causing the health checks to fail? I don't know, but I believe this is a new problem. I've been using this template for many weeks and didn't see anything like this earlier on. Perhaps it has something to do with the current AMI?

@andreaswittig I hope you have some ideas to help us. Our website cannot actually function without the custom permalinks (as it is required by BuddyPress).

@gerrymiller For a temporary workaround install this password plugin on your site. It will prevent the instances from crashing until we can find the reason for this. It works like htaccess authentication, but without actually using htaccess authentication (which is incompatible with CloudFront): https://wordpress.org/plugins/wp-secure-maintainance/

~ Michael

**Comment by gerrymiller** [comment]
@ILLUMINICE I may or may not have had the same problem. After setting Permalinks to Post Name (which sets up redirects via `.htaccess`), the health check will return a 301 instead of a 200. However, the template sets up health checks to only work on a 200 return, so they were failing after this change. I updated my template to accept both 200 and 301, and it works fine now.

**Comment by ILLUMINICE** [comment]
@gerrymiller Thanks for your responses both here and in the issue you had posted. That's interesting that this was a file ownership issue for you. I now think we're having similar but different problems.

My .htaccess file was always writable. In fact when I changed the permalink settings earlier I watched the .htaccess file get automatically changed, proving it's writable. The only difference is that my file ownership is set at apache:Beanstalk, since I need to give group access to Beanstalk for deployment. To my knowledge this hasn't at all impaired WordPress since both the owner and group permissions are identical.

I can actually tell you that the template does run a chown apache:apache when first installing WordPress. Just do a search in the template and you'll see it. I know this because it was one of the first things I had issues with (dealing with file ownership). I was originally using WinSCP to edit files directly, and couldn't do that with the apache:apache settings. Now that I'm using Beanstalk for deployment I don't need to alter files directly in WinSCP.

Regarding the HTTP statuses...

It appears we're having different results here, as I'm seeing a 403 Forbidden on the live server, but a 200 OK when checking from outside the server, which makes no sense to me.

I am certain however that the rewrites for the permalinks are involved in my particular problem, because when they are removed the health checks stabilize. There must be a connection between the HTTP statuses I'm seeing and my permalinks issue. I just don't know what yet...

If you have any ideas please feel free to share...

~ Michael

**Comment by ILLUMINICE** [comment]
@gerrymiller I did try just now changing my .htaccess file to apache:apache, instead of apache:Beanstalk to see if it would make any difference, and it appears not. My health checks are still failing and I still see that 403 Forbidden status. This is obviously why the health checks are failing, but I can't figure out why I have the 403 in the first place.

**Comment by ILLUMINICE** [comment]
@gerrymiller Ok, I'm hot on the trail of the cause now. I commented out all of my .htaccess code with the exception of the WordPress permalinks rewrites, and guess what, that 403 Forbidden has just changed to a 301 like you had.

This means there is something in my .htaccess config causing the 403. Now I just need to narrow down the exact cause.

After that I'll modify the CF template just as you did to accept 301 HTTP statuses for the health check. A 301 is just a permanent redirect, so this seems perfectly normal to me since that's how WP is implementing the permalinks.

**Comment by ILLUMINICE** [comment]
@gerrymiller @andreaswittig 

Ok, so here is the final determination...

What I had here was actually two separate problems, both of which were affecting the HTTP status codes.

The first issue was the 403 Forbidden code I was seeing. It turns out this was caused by special security code I had in my .htaccess file to prevent SQL injection attacks. It's a lot of long-winded code that I never vetted myself. I've been using it a long time and never knew it could cause an issue like this. I've now removed all of that code.

After this I was back to the identical issue that @gerrymiller had. My site shows a 200 OK if the permalinks are set to default, and a 301 if using custom permalinks. So my next step is to change the health check settings to allow 301 codes just as Gerry did.

Of course the random element in all of this for me was the password protection plugin I was using which shielded me from all of the health check failures. That's why I didn't know I had this problem for so long.

@gerrymiller Thank you for posting your solution, as it would have taken me forever to realize it was the permalink rewrites causing the 301.

I'm going to close this thread now since I know the final steps to solve everything. ;)

~ Michael
